gpii/node_modules/lifecycleManager/src/LifecycleManager.js

```diff
@@ -509,13 +534,43 @@ var gpii = fluid.registerNamespace("gpii");
       * @return {Promise} A promise for the action of restoring the system
       */
     gpii.lifecycleManager.restoreSystem = function (that, session, rootAction) {
-        var promises = fluid.transform(session.model.originalSettings, function (changesSolutionRecord, solutionId) {
-            var solutionRecord = gpii.lifecycleManager.transformSolutionSettings(changesSolutionRecord, gpii.settingsHandlers.changesToSettings);
-            return that.applySolution(solutionId, solutionRecord, session, [ "stop", "restore" ], rootAction);
+        var tasks = [];
+        fluid.each(session.model.originalSettings, function (changesSolutionRecord, solutionId) {
+            tasks.push(function () {
+                // check the current state of the solution to decide whether we should run the
+                // "restore", "update, or "stop"
+                return that.getSolutionRunningState(solutionId, changesSolutionRecord, session);
+            });
+            tasks.push(function () {
+                if (!fluid.isDestroyed(that)) { // See above comment for GPII-580
+                    // if solution is already running, call "update" directive - else use "start" directive
+                    var solutionRecord = gpii.lifecycleManager.transformSolutionSettings(changesSolutionRecord, gpii.settingsHandlers.changesToSettings);
+                    var recordedRunState = gpii.lifecycleManager.getSolutionRunningStateFromSnapshot(solutionRecord);
+                    var currentRunState = session.model.runningOnLogin[solutionId];
```

Comment:
It seems that doing this work before the "convert launchers to settings handlers" is making quite a mess